Fix oauth2 connector to support Ballerina version 0.95.3

Biruntha opened the pull request:
## Purpose
Keeping the Ballerina oauth2 Connector up-to Date with Ballerina changes.

## Goals
Fix oauth2 connector to support Ballerina version 0.95.2

## Approach
N/A

## User stories
N/A

## Release note
N/A

## Documentation
N/A

## Training
N/A

## Certification
N/A

## Marketing
N/A

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

## Samples
N/A

## Related PRs
N/A

## Migrations (if applicable)
N/A

## Test environment
Ballerina version 0.95.2
 
## Learning
N/A

Biruntha committed:
oauth2 connector added

Biruntha committed:
Pom file changes

anupama-pathirage commented on component/pom.xml:
```diff
@@ -0,0 +1,183 @@
+                        <!--</goals>-->
+                    <!--</execution>-->
+                <!--</executions>-->
+            <!--</plugin>-->
```
Shall we remove commented blocks here

anupama-pathirage commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+                        <!--<suiteXmlFile>src/test/resources/testng.xml</suiteXmlFile>-->
+                    <!--</suiteXmlFiles>-->
+                <!--</configuration>-->
+            <!--</plugin>-->
```
Shall we remove the commented blocks here

anupama-pathirage commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+<!--
+  ~ Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Wrong license year

anupama-pathirage commented on component/pom.xml:
```diff
@@ -0,0 +1,183 @@
+                        <!--</goals>-->
+                    <!--</execution>-->
+                <!--</executions>-->
+            <!--</plugin>-->
```
Shall we remove commented blocks here?

Biruntha replied:
Fixed

anupama-pathirage commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+<!--
+  ~ Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Incorrect license year

Biruntha replied:
Fixed

anupama-pathirage commented on pom.xml:
```diff
@@ -0,0 +1,350 @@
+                        <!--<suiteXmlFile>src/test/resources/testng.xml</suiteXmlFile>-->
+                    <!--</suiteXmlFiles>-->
+                <!--</configuration>-->
+            <!--</plugin>-->
```
Shall we remove commented blocks here?

Biruntha replied:
Fixed

Biruntha committed:
pom file changes

lafernando commented on component/src/main/java/connectors/Oauth2ConnectorExtensionPackageRepositoryProvider.java:
```diff
@@ -0,0 +1,40 @@
+ * @since 0.94
+ */
+@JavaSPIService("org.ballerinalang.spi.ExtensionPackageRepositoryProvider")
+public class Oauth2ConnectorExtensionPackageRepositoryProvider implements ExtensionPackageRepositoryProvider {
```
This should be "OAuth2...".

lafernando commented on component/src/org/wso2/ballerina/connectors/oauth2/ClientConnector.bal:
```diff
@@ -0,0 +1,173 @@
+    action get (string path, http:Request request) (http:Response, http:HttpConnectorError) {
+        http:Response response = {};
+
+        accessTokenValue = constructAuthHeader (request, accessTokenValue, accessToken);
```
Why are you taking in "accessTokenValue" and also returning "accessTokenValue". No one is using the returned value. And you don't have to pass it in also, since it's a service level variable which is accessible by the "constructAuthHeader". Also, the name of the function is wrong, it should be something like "populateAuthHeader", since you're passing in the "request", and you're putting the header value inside the function.

lafernando commented on component/src/org/wso2/ballerina/connectors/oauth2/ClientConnector.bal:
```diff
@@ -0,0 +1,173 @@
+        http:Request request = {};
+        request.setJsonPayload(originalPayload);
+
+        if ((response.getStatusCode() == 401) && (refreshToken != "" || refreshToken != "null")) {
```
These two lines are repeated in all the operations. That is a sign that it can be used as a common code to be used by several places. These two lines can be wrapped in to a boolean returning function like "checkAndRefreshToken", where it will check if the response is a 401, and then generates a new access token if so and returns it. We can either do a multi-return with a boolean and a string, or else, just return a string and check if it's empty/null. (null is supported in v0.95.3).

Biruntha committed:
Fix

Renamed to OAuth2ConnectorExtensionPackageRepositoryProvider
Fixed OAuth2 ClientConnector.bal file

lafernando commented on component/src/org/wso2/ballerina/connectors/oauth2/ClientConnector.bal:
```diff
@@ -0,0 +1,172 @@
+}
+
+function populateAuthHeader (http:Request request, string accessToken) {
+    if (accessTokenValue == "") {
```
If you put "|| accessTokenValue == null", this code will also work with Ballerina 0.95.3, since from there onwards the default value of string would be null.

lafernando replied:
On second thought, let's just make it work from 0.95.3 onwards. So basically just check for == null.

Biruntha committed:
Fixed

Check null values using '== null'

Biruntha committed:
Connector version changed
